File: __java-sample/src/main/java/bytecode/Singleton.java
package bytecode;

/**
 * Created by IntelliJ IDEA.
 *
 * @author Eric Cui
 * @date 2017/12/20 11:02
 * Desc    Setting | Editor | File and Code Templates
 */
public class Singleton {
    private static Singleton instance = new Singleton();
    public static int counter1;
    public static int counter2 = 0;
    private Singleton() {
        counter1++;
        counter2++;
    }
    public static Singleton getInstance() {
        return instance;
    }

    public static void main(String[] args) {
        Singleton instance = Singleton.getInstance();
        System.out.println("counter1 = " + Singleton.counter1 + ", counter2 = " + Singleton.counter2);
    }
}

File: kotlin-sample/src/singleton/Singleton.kt
package singleton

/**
 * Created by IntelliJ IDEA.
 * @author Eric
 * Date    2017/9/1 10:01
 * Desc    Setting | Editor | File and Code Templates
 *
 */
class Singleton private constructor() {
    companion object {
        val instance: Singleton by lazy { Singleton() }
    }

    fun getValue(key: String): String {
        return key + ":val"
    }
}

fun main(args: Array<String>) {
    // Kotlin 中调用
    val key = "Hello"
    Singleton.instance.getValue(key)
}
